File: crates/everyaios-ipc/src/message.rs
//! JSON-RPC 2.0 message types used on the wire.
//!
//! A minimal, strict-enough subset: `Request` (with optional `id` → a
//! notification when absent), `Response` (result or error). Field-level
//! validation happens in the app layer; this crate only carries the shape.

use serde::{Deserialize, Serialize};

/// JSON-RPC 2.0 marker — every message declares this.
pub const JSONRPC: &str = "2.0";

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Request {
    pub jsonrpc: String,
    /// Method name (e.g. `chat/stream`, `browser/act`, `vault/rotate`).
    pub method: String,
    /// Positional or named params. `None` = absent.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub params: Option<serde_json::Value>,
    /// Present = a request awaiting a response; absent = notification.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub id: Option<serde_json::Value>,
}

impl Request {
    pub fn new(method: impl Into<String>) -> Self {
        Self {
            jsonrpc: JSONRPC.into(),
            method: method.into(),
            params: None,
            id: Some(serde_json::json!(next_id())),
        }
    }

    pub fn with_params(mut self, params: serde_json::Value) -> Self {
        self.params = Some(params);
        self
    }

    /// True when this is a notification (no `id`) — fire-and-forget.
    pub fn is_notification(&self) -> bool {
        self.id.is_none()
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Response {
    pub jsonrpc: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub id: Option<serde_json::Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub result: Option<serde_json::Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<JsonRpcError>,
}

impl Response {
    pub fn ok(id: serde_json::Value, result: serde_json::Value) -> Self {
        Self {
            jsonrpc: JSONRPC.into(),
            id: Some(id),
            result: Some(result),
            error: None,
        }
    }

    pub fn err(id: serde_json::Value, code: i64, message: impl Into<String>) -> Self {
        Self {
            jsonrpc: JSONRPC.into(),
            id: Some(id),
            result: None,
            error: Some(JsonRpcError {
                code,
                message: message.into(),
                data: None,
            }),
        }
    }
}

/// Standard JSON-RPC error object (code/message/data).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct JsonRpcError {
    pub code: i64,
    pub message: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub data: Option<serde_json::Value>,
}

impl JsonRpcError {
    pub const PARSE_ERROR: i64 = -32700;
    pub const INVALID_REQUEST: i64 = -32600;
    pub const METHOD_NOT_FOUND: i64 = -32601;
    pub const INTERNAL_ERROR: i64 = -32603;

    pub fn new(code: i64, message: impl Into<String>) -> Self {
        Self {
            code,
            message: message.into(),
            data: None,
        }
    }

    pub fn method_not_found(method: &str) -> Self {
        Self::new(
            Self::METHOD_NOT_FOUND,
            format!("method not found: {method}"),
        )
    }
}

/// Monotonic id generator for requests (simple counter; the real sidecar
/// will use a per-session counter seeded from the handshake).
static ID_COUNTER: std::sync::atomic::AtomicU64 = std::sync::atomic::AtomicU64::new(1);

fn next_id() -> u64 {
    ID_COUNTER.fetch_add(1, std::sync::atomic::Ordering::Relaxed)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn request_serializes_with_id() {
        let req = Request::new("browser/snapshot")
            .with_params(serde_json::json!({"mode": "interactive"}));
        let json = serde_json::to_string(&req).unwrap();
        assert!(json.contains("\"jsonrpc\":\"2.0\""));
        assert!(json.contains("\"method\":\"browser/snapshot\""));
        assert!(json.contains("\"id\":"));
        assert!(json.contains("\"params\":{\"mode\":\"interactive\"}"));
    }

    #[test]
    fn request_without_id_is_notification() {
        let mut req = Request::new("session/ping");
        req.id = None;
        assert!(req.is_notification());
        let json = serde_json::to_string(&req).unwrap();
        assert!(!json.contains("\"id\""));
    }

    #[test]
    fn response_ok_and_err_roundtrip() {
        let ok = Response::ok(serde_json::json!(1), serde_json::json!({"ok": true}));
        let back: Response = serde_json::from_str(&serde_json::to_string(&ok).unwrap()).unwrap();
        assert_eq!(back, ok);
        assert!(back.error.is_none());

        let err = Response::err(
            serde_json::json!(1),
            JsonRpcError::METHOD_NOT_FOUND,
            "browser/nope",
        );
        let back: Response = serde_json::from_str(&serde_json::to_string(&err).unwrap()).unwrap();
        assert_eq!(back.error.unwrap().code, JsonRpcError::METHOD_NOT_FOUND);
    }

    #[test]
    fn method_not_found_helper() {
        let e = JsonRpcError::method_not_found("browser/nope");
        assert_eq!(e.code, -32601);
        assert!(e.message.contains("browser/nope"));
    }
}

File: packages/coordinator/src/message.ts
/**
 * EveryAIOS IPC JSON-RPC message types — the TS mirror of
 * `everyaios-ipc/src/message.rs`. Wire shape is camelCase and must match
 * the Rust `serde(rename_all = "camelCase")` structs byte-for-byte.
 */

/** JSON-RPC 2.0 marker — every message declares this. */
export const JSONRPC = "2.0";

/** Standard JSON-RPC error codes (mirror `JsonRpcError` constants in Rust). */
export const ERROR_CODES = {
  PARSE_ERROR: -32700,
  INVALID_REQUEST: -32600,
  METHOD_NOT_FOUND: -32601,
  INTERNAL_ERROR: -32603,
} as const;

/** A JSON-RPC request or notification (no `id` = notification). */
export interface Request {
  jsonrpc: string;
  /** Method name (e.g. `chat/stream`, `browser/act`, `vault/rotate`). */
  method: string;
  /** Positional or named params. Omitted when absent. */
  params?: unknown;
  /** Present = a request awaiting a response; absent = notification.
   *  Explicit `null` is allowed (JSON-RPC 2.0 §2.2; Rust: `Some(Null)`). */
  id?: string | number | null;
}

/** A JSON-RPC error object (code/message/data). */
export interface JsonRpcError {
  code: number;
  message: string;
  data?: unknown;
}

/** A JSON-RPC response (either `result` or `error`, never both). */
export interface Response {
  jsonrpc: string;
  id: string | number | null;
  result?: unknown;
  error?: JsonRpcError;
}

/**
 * Guards: is this value a well-formed request object?
 *
 * `id` may be a string, number, or explicit `null` (JSON-RPC 2.0 §2.2 allows
 * null; mirrors Rust `Option<Value>` where `Some(Null)` is a request, not a
 * notification). Only an *absent* `id` means notification — matching Rust's
 * `is_notification()` (`id.is_none()`).
 */
export function isRequest(value: unknown): value is Request {
  if (typeof value !== "object" || value === null) return false;
  const v = value as Record<string, unknown>;
  return (
    v.jsonrpc === JSONRPC &&
    typeof v.method === "string" &&
    (v.id === undefined || v.id === null || typeof v.id === "string" || typeof v.id === "number")
  );
}

/** Build a success response. */
export function ok(id: string | number | null, result: unknown): Response {
  return { jsonrpc: JSONRPC, id, result };
}

/** Build an error response. */
export function err(id: string | number | null, code: number, message: string): Response {
  return { jsonrpc: JSONRPC, id, error: { code, message } };
}

/** Convenience: METHOD_NOT_FOUND response. */
export function methodNotFound(id: string | number | null, method: string): Response {
  return err(id, ERROR_CODES.METHOD_NOT_FOUND, `method not found: ${method}`);
}
